From 5f122457139afa35dd8eac40a9fb823cd3494e22 Mon Sep 17 00:00:00 2001
From: Jan Engelhardt <jengelh@inai.de>
Date: Thu, 3 Sep 2015 02:29:37 +0200
Subject: [PATCH] Handle .deb files that have sysv ar(1) filenames

---
 Build/Deb.pm | 20 ++++++++++++--------
 1 file changed, 12 insertions(+), 8 deletions(-)

--- a/Build/Deb.pm
+++ b/Build/Deb.pm
@@ -193,11 +193,12 @@ sub debq {
   my $data = '';
   sysread(DEBF, $data, 4096);
   if (length($data) < 8+60) {
-    warn("$fn: not a debian package\n");
+    warn("$fn: not a debian package - header too short\n");
     close DEBF unless ref $fn;
     return ();
   }
-  if (substr($data, 0, 8+16) ne "!<arch>\ndebian-binary   ") {
+  if (substr($data, 0, 8+16) ne "!<arch>\ndebian-binary   " &&
+      substr($data, 0, 8+16) ne "!<arch>\ndebian-binary/  ") {
     close DEBF unless ref $fn;
     return ();
   }
@@ -213,7 +214,8 @@ sub debq {
     }
   }
   $data = substr($data, 8 + 60 + $len);
-  if (substr($data, 0, 16) ne 'control.tar.gz  ') {
+  if (substr($data, 0, 16) ne 'control.tar.gz  ' &&
+      substr($data, 0, 16) ne 'control.tar.gz/ ') {
     warn("$fn: control.tar.gz is not second ar entry\n");
     close DEBF unless ref $fn;
     return ();
@@ -249,7 +251,7 @@ sub debq {
       warn("$fn: corrupt control.tar.gz file\n");
       return ();
     }
-    if ($n eq './control') {
+    if ($n eq './control' || $n eq "control") {
       $control = substr($data, 512, $len);
       last;
     }
@@ -324,12 +326,13 @@ sub queryhdrmd5 {
   my $data = '';
   sysread(F, $data, 4096);
   if (length($data) < 8+60) {
-    warn("$bin: not a debian package\n");
+    warn("$bin: not a debian package - header too short\n");
     close F;
     return undef;
   }
-  if (substr($data, 0, 8+16) ne "!<arch>\ndebian-binary   ") {
-    warn("$bin: not a debian package\n");
+  if (substr($data, 0, 8+16) ne "!<arch>\ndebian-binary   " &&
+      substr($data, 0, 8+16) ne "!<arch>\ndebian-binary/  ") {
+    warn("$bin: not a debian package - no \"debian-binary\" entry\n");
     close F;
     return undef;
   }
@@ -345,7 +348,8 @@ sub queryhdrmd5 {
     }
   }
   $data = substr($data, 8 + 60 + $len);
-  if (substr($data, 0, 16) ne 'control.tar.gz  ') {
+  if (substr($data, 0, 16) ne 'control.tar.gz  ' &&
+      substr($data, 0, 16) ne 'control.tar.gz/ ') {
     warn("$bin: control.tar.gz is not second ar entry\n");
     close F;
     return undef;
